futures-cpupool support with boxed future

The code below cannot be compiled. That seems to be because `Box<Future>` is not `Send`. While `#[async(boxed)]` is not necessary and `#[async]` works well instead here, it's better to support some kind of boxed generator which is `Send` as they would be common in complex applications (e.g. when writing recursive functions.)

This issue may be related to #15 because the problem is caused by trait objects. I guess something like `Box<G> where G: Generator` fixes it. Or some way to support `Box<Future + Send>`?

```rust
#![feature(proc_macro, generators)]

extern crate futures_await as futures;
extern crate futures_cpupool;

use futures::prelude::*;
use futures_cpupool::*;

#[async(boxed)]
fn foo() -> Result<(), ()> {
    Ok(())
}

fn main() {
    CpuPool::new_num_cpus().spawn(foo()).wait();
}
```

Build log:

```
> cargo build
   Compiling rust-async-await-examples v0.1.0 (file:///home/raviqqe/src/github.com/raviqqe/rust-async-await-examples)
error[E0277]: the trait bound `futures::Future<Error=(), Item=()>: std::marker::Send` is not satisfied
  --> src/main.rs:15:29
   |
15 |     CpuPool::new_num_cpus().spawn(foo()).wait();
   |                             ^^^^^ `futures::Future<Error=(), Item=()>` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `futures::Future<Error=(), Item=()>`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<futures::Future<Error=(), Item=()>>`
   = note: required because it appears within the type `std::boxed::Box<futures::Future<Error=(), Item=()>>`

error: aborting due to previous error

error: Could not compile `rust-async-await-examples`.

To learn more, run the command again with --verbose.
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -24,5 +24,6 @@ futures-await-await-macro = { path = "futures-await-await-macro", version = "0.1
 futures = "0.1"
 
 [dev-dependencies]
+futures-cpupool = "0.1"
 tokio-core = "0.1"
 tokio-io = "0.1"

--- a/futures-await-async-macro/src/lib.rs
+++ b/futures-await-async-macro/src/lib.rs
@@ -225,25 +225,30 @@ where F: FnOnce(&Type) -> proc_macro2::TokenStream
 #[proc_macro_attribute]
 pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
     // Handle arguments to the #[async] attribute, if any
-    let attribute = attribute.to_string();
-    let boxed = if attribute == "( boxed )" {
-        true
-    } else if attribute == "" {
-        false
-    } else {
-        panic!("the #[async] attribute currently only takes `boxed` as an arg");
+    let (boxed, send) = match &attribute.to_string() as &str {
+        "( boxed )" => (true, false),
+        "( boxed_send )" => (true, true),
+        "" => (false, false),
+        _ => panic!("the #[async] attribute currently only takes `boxed` as an arg"),
     };
 
     async_inner(boxed, function, quote_cs! { ::futures::__rt::gen }, |output| {
         // TODO: can we lift the restriction that `futures` must be at the root of
         //       the crate?
         let output_span = first_last(&output);
-        let return_ty = if boxed {
+        let return_ty = if boxed && !send {
             quote_cs! {
                 ::futures::__rt::std::boxed::Box<::futures::Future<
                     Item = <! as ::futures::__rt::IsResult>::Ok,
                     Error = <! as ::futures::__rt::IsResult>::Err,
                 >>
+            }
+        } else if boxed && send {
+            quote_cs! {
+                ::futures::__rt::std::boxed::Box<::futures::Future<
+                    Item = <! as ::futures::__rt::IsResult>::Ok,
+                    Error = <! as ::futures::__rt::IsResult>::Err,
+                > + Send>
             }
         } else {
             // Dunno why this is buggy, hits weird typecheck errors in tests

--- a/tests/smoke.rs
+++ b/tests/smoke.rs
@@ -6,10 +6,12 @@
 #![feature(proc_macro, conservative_impl_trait, generators)]
 
 extern crate futures_await as futures;
+extern crate futures_cpupool;
 
 use std::io;
 
 use futures::prelude::*;
+use futures_cpupool::CpuPool;
 
 #[async]
 fn foo() -> Result<i32, i32> {
@@ -56,6 +58,11 @@ fn _foo7<T>(t: T) -> Result<T, i32>
 #[async(boxed)]
 fn _foo8(a: i32, b: i32) -> Result<i32, i32> {
     return Ok(a + b)
+}
+
+#[async(boxed_send)]
+fn _foo9() -> Result<(), ()> {
+    Ok(())
 }
 
 #[async]
@@ -233,4 +240,10 @@ fn poll_stream_after_error() {
     assert_eq!(s.poll(), Ok(Async::Ready(Some(5))));
     assert_eq!(s.poll(), Err(()));
     assert_eq!(s.poll(), Ok(Async::Ready(None)));
+}
+
+#[test]
+fn run_boxed_future_in_cpu_pool() {
+    let pool = CpuPool::new_num_cpus();
+    pool.spawn(_foo9()).wait().unwrap();
 }

